### ISSUE TITLE
Remove GlobalOpenTelemetry.set call from GlobalOpenTelemetry.get?

### ISSUE THREAD

**Issue by jeanbisutti** [open-telemetry/opentelemetry-java#5555]
Today, [`GlobalOpenTelemetry.get` calls `set(OpenTelemetry.noop())`](https://github.com/open-telemetry/opentelemetry-java/blob/8f1a7b1c95e8d4a6591be7b48e54fe89da4f386b/api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java#L82) if no OTel autoconfiguration is found or if `GlobalOpenTelemetry.set` was not previously called. Maybe the `set(OpenTelemetry.noop())`  call could be removed and just return `OpenTelemetry.noop()`. Maybe [it could help avoid this type of problem](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/8773), which prevents the application from starting up.

**Comment by jack-berg** [comment]
Check out this [comment](https://github.com/open-telemetry/opentelemetry-java/pull/5010#discussion_r1039020089) and the broader discussion in that PR.

TL;DR is that I'm fairly convinced that the current behavior is indeed the right behavior. Calling `set(OpenTelemetry.noop())` ensures that all calls to `GlobalOpenTelemetry.get()` get the same response, regardless of when they call it in the application lifecycle. This invariant simplifies debugging, even if its inconvenient the first time you run across it.

Options for a user that runs across an issue with this:
- Opt into autoconfigure via `otel.java.global-autoconfigure.enabled=true`
- Eliminate calls to `GlobalOpenTelemetry.get()`, which we discourage anyway
- Adjust application startup so that OpenTelemetrySdk setup happens before anything else

The instrumentation issue you link to is an interesting one because we currently break out own advice regarding avoiding `GlobalOpenTelemetry.get()` calls in the log appenders. We recently talked about this in the log SIG and are investigating programatic configuration of log appender OpenTelemetry instances. If we don't run into any issues, we can remove the `GlobalOpenTelemetry.get()` call (or make it opt-in), instead recommending programatic configuration.

**Comment by gabrieljones** [comment]
> Eliminate calls to GlobalOpenTelemetry.get(), which we discourage anyway

Is there more detail on this somewhere?

**Comment by jack-berg** [comment]
There's [javadoc](https://github.com/open-telemetry/opentelemetry-java/blob/4fdfc15633acc1e508564a23b6e17800ed4a1547/api/all/src/main/java/io/opentelemetry/api/GlobalOpenTelemetry.java#L34-L37) on the GlobalOpenTelemetry class. 

There's some additional notes [here](https://github.com/open-telemetry/opentelemetry-java/blob/main/docs/sdk-configuration.md#the-global-instance-of-the-sdk), but that's an old document so some of it may be outdated.

**Comment by gabrieljones** [comment]
I am running into a problem here.
I do `AutoConfiguredOpenTelemetrySdk.builder().build()` **after** the app loads the config from a url.

That config contains all of my otel properties that I set at runtime via `System.setProperty`.
This config load emits logs before the otel config is set and then I run into this problem.

**Comment by jack-berg** [comment]
Can you adjust whatever code is calling `GlobalOpenTelemetry.get()` to use `AutoConfiguredOpenTelemetrySdk.builder().build()` instead?

**Comment by gabrieljones** [comment]
It is `OpenTelemetryAppender` that is calling `GlobalOpenTelemetry.get()`

**Comment by gabrieljones** [comment]
Also regarding:
> Libraries that want to export telemetry data using OpenTelemetry MUST only depend on the opentelemetry-api package and should never configure or depend on the OpenTelemetry SDK.
> --https://opentelemetry.io/docs/instrumentation/java/manual/

How should a library get a reference to an `OpenTelemetry` instance?

**Comment by jack-berg** [comment]
> It is io.opentelemetry.instrumentation.log4j.appender.v2_17.OpenTelemetryAppender that is calling GlobalOpenTelemetry.get()

Ah yes. This is problematic because the log appenders currently break our own advice. This was talked about in detail in the 6/22 java SIG and we're working towards a solution. I proposed one possible solution [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8791), which advocates for programatic configuration of the `OpenTelemetry` instance with the log appenders, and changes the `GlobalOpenTelemetry` behavior to opt-in.

> How should a library get a reference to an OpenTelemetry instance?

Generally, the library should provide some sort of configuration API to accept an `OpenTelemetry` as an argument during initialization. Log appenders are a particularly tricky situation because: 

1. The config is typically (but not required to be) XML based, which doesn't allow for an OpenTelemetry configured elsewhere to be injected
2. The appenders are typically initialized very early in the app lifecycle before the user has the chance to configure an OpenTelemetrySdk

This is definitely a problem in 1.27.0, since this is the first version in which GlobalLoggerProvider was replaced with GlobalOpenTelemetry, which is more strict. We're working on figuring out a solution that is more user friendly. [opentelemetry-java-instrumentation#9773](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/8773) is the right issue to follow.
